CRM/Contribute/BAO/Contribution.php

```diff
@@ -200,6 +196,14 @@ public static function add(&$params) {
     }
 
     $result = $contribution->save();
+    // Save invoice number if appropriate. Note we used to try to
+    // avoid a second save but check https://lab.civicrm.org/dev/core/-/issues/5004
+    // to see how that worked out....
+    if (!$contributionID) {
+      if (empty($params['invoice_number']) && \Civi::settings()->get('invoicing')) {
+        $contribution->invoice_number = self::getInvoiceNumber($contribution->id);
+      }
+    }
```

Comment:
@eileenmcnaughton I don't see where this would actually write the invoice number to the contribution. Also I wonder if it would be sufficient to either (a) move this code to the line before `$contribution->save()` as there's a lot of "work"(delay) going on between the original code and the save. or (b) move the code to just before save and wrap in a lock.

Reply:
@mattwire that's because I didn't push up the lastest version of the commit....

Reply:
I think there would always be some leakage on the line before - I'm not sure about a lock - it hurts my head a bit to think through that - but maybe after coffee

Reply:
I think there would always be some leakage on the line before  - I'm not sure about a lock - it hurts my head a bit to think through that - but maybe after coffee

Reply:
@mattwire so I think what we are dealing with here is that the auto increment may not match the collace so waiting until we have a known id for the new the new contribution so I don't think it can go before the save as we don't get the assigned id until the save() is called first